src/rdkafka_cgrp.c

```diff
@@ -2807,17 +2862,53 @@ void rd_kafka_cgrp_handle_ConsumerGroupHeartbeat(rd_kafka_t *rk,
                         }
 
-                        rd_rkb_dbg(rkb, CGRP, "HEARTBEAT",
-                                   "Heartbeat response received target "
-                                   "assignment \"%s\"",
-                                   assigned_topic_partitions_str);
+                        rd_rkb_dbg(
+                            rkb, CGRP, "HEARTBEAT",
+                            "ConsumerGroupHeartbeat response received target "
+                            "assignment \"%s\"",
+                            assigned_topic_partitions_str);
                 }
 
                 if (assigned_topic_partitions) {
+                        rd_bool_t assignment_updated = rd_true;
                         RD_IF_FREE(rkcg->rkcg_next_target_assignment,
                                    rd_kafka_topic_partition_list_destroy);
-                        rkcg->rkcg_next_target_assignment =
-                            assigned_topic_partitions;
+                        rkcg->rkcg_next_target_assignment = NULL;
+                        if (rkcg->rkcg_target_assignment) {
+                                if (!rd_kafka_topic_partition_list_cmp(
+                                        assigned_topic_partitions,
+                                        rkcg->rkcg_target_assignment,
+                                        rd_kafka_topic_partition_by_id_cmp)) {
+                                        /* If target assignment is present and
+                                         * the new assignment is same as target
+                                         * assignment, then we are already in
+                                         * process of adding that target
+                                         * assignment. We can ignore this new
+                                         * assignment.*/
+                                        assignment_updated = rd_false;
+                                }
+                        } else if (rkcg->rkcg_current_assignment) {
+                                if (!rd_kafka_topic_partition_list_cmp(
+                                        assigned_topic_partitions,
+                                        rkcg->rkcg_current_assignment,
+                                        rd_kafka_topic_partition_by_id_cmp)) {
+                                        /* If target assignment is not present
+                                         * then if the current assignment is
+                                         * present and the new assignment is
+                                         * same as current assignment, then we
+                                         * are already at correct assignment. We
+                                         * can ignore this new assignment.*/
+                                        assignment_updated = rd_false;
+                                }
+                        }
+                        if (assignment_updated) {
```

Comment:
Can introduce this function and use it here and in `rd_kafka_cgrp_consumer_handle_next_assignment`

```c
static rd_bool_t rd_kafka_cgrp_consumer_is_assignment_different(
    rd_kafka_cgrp_t *rkcg,
    rd_kafka_topic_partition_list_t *new_target_assignment) {
        rd_bool_t is_assignment_different;
        if (rkcg->rkcg_target_assignment) {
                is_assignment_different = rd_kafka_topic_partition_list_cmp(
                    new_target_assignment, rkcg->rkcg_target_assignment,
                    rd_kafka_topic_partition_by_id_cmp);
        } else {
                is_assignment_different = rd_kafka_topic_partition_list_cmp(
                    new_target_assignment, rkcg->rkcg_current_assignment,
                    rd_kafka_topic_partition_by_id_cmp);
        }
        return is_assignment_different;
}
``` 
```suggestion
                        if (rd_kafka_cgrp_consumer_is_assignment_different(rkcg, assigned_topic_partitions)) {
```

Reply:
Extracting the function.

We should not directly return response from `rd_kafka_topic_partition_list_cmp` as `rd_bool_t`. `rd_kafka_topic_partition_list_cmp` returns `int` and `rd_bool_t` is 1 byte. Updated the function to cater this.